### PR TITLE
Ros 256 - First of several PRs. This one is tuple-cons primitive

### PR DESCRIPTION
**Pull request by kjmcqueen**
https://rchain.atlassian.net/browse/ROS-256

This is the first PR related to this Jira ticket.

Primitive: tuple-cons

### CODE REVIEW

**Comment by tymm** [comment]
Travis says that the build failed. Please make sure that it compiles.

**Comment by kjmcqueen** [comment]
I saw that the build failed and looked at it.  It doesn't have anything to do with my PR.  I believe this is a pre-existing condition.
{quote}
[info] Compiling 39 Scala sources to /home/travis/build/rchain/rchain/roscala/target/scala-2.12/classes ...
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala:490:9: match may not be exhaustive.
[warn] It would fail on the following inputs: Left(Absent), Left(Invalid), Left(PrimErrorWrapper(_)), Left(RuntimeError(_)), Left(Suspend), Left(Upcall)
[warn]         result match {
[warn]         ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala:529:9: match may not be exhaustive.
[warn] It would fail on the following inputs: Left(Absent), Left(Invalid), Left(PrimErrorWrapper(_)), Left(RuntimeError(_)), Left(Suspend), Left(Upcall)
[warn]         result match {
[warn]         ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala:561:9: match may not be exhaustive.
[warn] It would fail on the following inputs: Left(Absent), Left(Invalid), Left(PrimErrorWrapper(_)), Left(RuntimeError(_)), Left(Suspend), Left(Upcall)
[warn]         result match {
[warn]         ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala:589:9: match may not be exhaustive.
[warn] It would fail on the following inputs: Left(Absent), Left(Invalid), Left(PrimErrorWrapper(_)), Left(RuntimeError(_)), Left(Suspend), Left(Upcall)
[warn]         result match {
[warn]         ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala:684:5: match may not be exhaustive.
[warn] It would fail on the following inputs: Left(DeadThread), Left(Invalid), Left(PrimErrorWrapper(_)), Left(RuntimeError(_)), Left(Suspend)
[warn]     value match {
[warn]     ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala:704:5: match may not be exhaustive.
[warn] It would fail on the following inputs: Left(DeadThread), Left(Invalid), Left(PrimErrorWrapper(_)), Left(RuntimeError(_)), Left(Suspend)
[warn]     value match {
[warn]     ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:23:59: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFixnum(0)) {
[warn]                                                           ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:59:59: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFixnum(1)) {
[warn]                                                           ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:205:70: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFixnum(Int.MaxValue)) {
[warn]                                                                      ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:222:70: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFixnum(Int.MinValue)) {
[warn]                                                                      ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:324:60: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFixnum(~0)) {
[warn]                                                            ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:340:59: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFixnum(0)) {
[warn]                                                           ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:356:59: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFixnum(0)) {
[warn]                                                           ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/Fixnum.scala:386:71: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       val commonBits = ctxt.argvec.elem.take(n).foldLeft(RFixnum(~0)) {
[warn]                                                                       ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/RblFloat.scala:28:60: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFloat(0.0)) {
[warn]                                                            ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/RblFloat.scala:64:58: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFloat(1)) {
[warn]                                                          ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/RblFloat.scala:190:72: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFloat(Double.MaxValue)) {
[warn]                                                                        ^
[warn] /home/travis/build/rchain/rchain/roscala/src/main/scala/coop/rchain/rosette/prim/RblFloat.scala:207:72: match may not be exhaustive.
[warn] It would fail on the following input: (_, _)
[warn]       Right(ctxt.argvec.elem.take(n).foldLeft(RFloat(Double.MinValue)) {
[warn]                                                                        ^
[warn] 18 warnings found
[info] Done compiling.
[info] Formatting 9 Scala sources in root:test ...
[warn] /home/travis/build/rchain/rchain/roscala/src/test/scala/coop/rchain/rosette/prim/FixnumSpec.scala:23: error: illegal start of simple expression
[warn] )
[warn]   ^
[warn] /home/travis/build/rchain/rchain/roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala:23: error: illegal start of simple expression
[warn] )
[warn]   ^
[info] Compiling 9 Scala sources to /home/travis/build/rchain/rchain/roscala/target/scala-2.12/test-classes ...
[warn] /home/travis/build/rchain/rchain/roscala/src/test/scala/coop/rchain/rosette/prim/FixnumSpec.scala:4:71: imported `Tuple' is permanently hidden by definition of object Tuple in package prim
[warn] import coop.rchain.rosette.{Ctxt, Fixnum => RFixnum, Ob, PC, RblBool, Tuple}
[warn]                                                                       ^
[error] /home/travis/build/rchain/rchain/roscala/src/test/scala/coop/rchain/rosette/prim/FixnumSpec.scala:15:19: coop.rchain.rosette.prim.Tuple.type does not take parameters
[error]     argvec = Tuple(1, RFixnum(1)),
[error]                   ^
[warn] /home/travis/build/rchain/rchain/roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala:4:91: imported `Tuple' is permanently hidden by definition of object Tuple in package prim
[warn] import coop.rchain.rosette.{Ctxt, Fixnum => RFixnum, Ob, PC, RblBool, RblFloat => RFloat, Tuple}
[warn]                                                                                           ^
[error] /home/travis/build/rchain/rchain/roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala:15:19: coop.rchain.rosette.prim.Tuple.type does not take parameters
[error]     argvec = Tuple(1, RFixnum(1)),
[error]                   ^
[warn] /home/travis/build/rchain/rchain/roscala/src/test/scala/coop/rchain/rosette/utils/opcodes/Dsl.scala:26:18: non-variable type argument shapeless.Witness in type pattern List[shapeless.Witness] (the underlying of List[shapeless.Witness]) is unchecked since it is eliminated by erasure
[warn]         case (l: List[Witness], r) =>
[warn]                  ^
[warn] three warnings found
[error] two errors found
[error] (root/test:compileIncremental) Compilation failed
[error] Total time: 151 s, completed Feb 1, 2018 8:49:24 PM
{quote}

**Comment by pyrocto** [comment]
In the test that's failing, your new Tuple object is shadowing the other one.

**Comment by kjmcqueen** [comment]
You're talking about the one in ...rosette/Tuple.scala vs mine in ...rosette/prim/Tuple.scala?  I can see that.  I guess I assumed (incorrectly?) that they would somehow "merge".

I'm not sure why it doesn't fail on my local machine, or what to do to fix it.

I guess my options are:

1. Move my code into ...rosette/Tuple.scala
2. Move the Tuple object from ...rosette/Tuple.scala into ...rosette/prim/Tuple.scala
3. Rename one or the other of the objects

I'm trying to follow the pattern laid out in https://rchain.atlassian.net/wiki/spaces/ROS/pages/92733451/Translating+Rosette+Primitives
 but the example there doesn't show where the object is to be defined.  I followed the example of  ...rosette/Fixnum.scala and ...rosette/prim/Fixnum.scala, but that example does not have a Fixnum object define in the top level scala module.

BTW, I feel it is confusing to have modules named the same even if they are in different directories.

The bottom line is that I think I understand the problem, I'm just looking for guidance from the team as how to resolve it.

**Comment by pyrocto** [comment]
If you want your code to be part of the Tuple object, it should be in the file where the Tuple object is defined, and the place for a class object is in the same file as the class.  If you don't want your code to be part of the Tuple object, you need a different name.  So 1 or 3, but I'll leave it to @tymm to tell you whether it should be part of the Tuple object or not.

**Comment by tymm** [comment]
@kjmcqueen my fault. Just rename `prim/Tuple.scala` to `prim/tuple.scala` and the `object` inside of `prim/tuple.scala` to `tuple` instead of `Tuple`. This should resolve the issues and is also more in line with the naming conventions (see: https://docs.scala-lang.org/style/naming-conventions.html#objects).

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package coop.rchain.rosette.prim
+
+import coop.rchain.rosette.{
+  Ctxt,
+  Fixnum => RFixnum,
+  Ob,
+  RblBool,
+  Tuple => RTuple
+}
+import coop.rchain.rosette.macros.{checkArgumentMismatch, checkTypeMismatch}
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: Seems like you don't need the `checkTypeMismatch` import.

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package coop.rchain.rosette.prim
+
+import coop.rchain.rosette.{
+  Ctxt,
+  Fixnum => RFixnum,
+  Ob,
+  RblBool,
+  Tuple => RTuple
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: With `object tuple` we don't need the renaming anymore.

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package coop.rchain.rosette.prim
+
+import coop.rchain.rosette.{
+  Ctxt,
+  Fixnum => RFixnum,
+  Ob,
+  RblBool,
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: You don't need that here.

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import coop.rchain.rosette.macros.{checkArgumentMismatch, checkTypeMismatch}
+import coop.rchain.rosette.prim.Prim._
+
+object Tuple {
+  object tplCons extends Prim {
+    override val name: String = "tuple-cons"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    //@checkTypeMismatch[RTuple]    arg0 can be any Ob
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: You can remove that.

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      checkRTuple(1, elem).map { tup => // arg1 must be a Tuple
+        RTuple.cons(elem(0), tup)
+      }
+    }
+
+  }
+
+  private def checkRTuple(n: Int, elem: Seq[Ob]): Either[PrimError, RTuple] =
+    if (!elem(n).isInstanceOf[RTuple]) {
+      Left(TypeMismatch(n, RTuple.getClass().getName()))
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: It's convention to use `RTuple.getClass.getName` here since `getClass` and `getName` have no side effects. At least that's what IntelliJ is telling me ;)

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  object tplCons extends Prim {
+    override val name: String = "tuple-cons"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    //@checkTypeMismatch[RTuple]    arg0 can be any Ob
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RTuple] = {
+      val elem = ctxt.argvec.elem
+      checkRTuple(1, elem).map { tup => // arg1 must be a Tuple
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: `checkRTuple(1, elem).map(RTuple.cons(elem.head, _))`

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala (hunk 3 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: I'll take out the rename, but leave Fixnum because I'll need it later.

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala (hunk 1 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: Leaving it because I'll need it later

File: roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala (hunk 5 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/Tuple.scala]: Sorry, my C/C++ roots are showing...

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    override val name: String = "tuple-cons"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Tuple] = {
+      val elem = ctxt.argvec.elem
+      checkTuple(1, elem).map(Tuple.cons(elem.head, _)) // arg1 must be a Tuple
+    }
+
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Please remove newline.

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  private def checkTuple(n: Int, elem: Seq[Ob]): Either[PrimError, Tuple] =
+    if (!elem(n).isInstanceOf[Tuple]) {
+      Left(TypeMismatch(n, Tuple.getClass().getName()))
+    } else {
+      Right(elem(n).asInstanceOf[Tuple])
+    }
+
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Please remove newline.

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    override def fn(ctxt: Ctxt): Either[PrimError, Tuple] = {
+      val elem = ctxt.argvec.elem
+      checkTuple(1, elem).map(Tuple.cons(elem.head, _)) // arg1 must be a Tuple
+    }
+
+  }
+
+  private def checkTuple(n: Int, elem: Seq[Ob]): Either[PrimError, Tuple] =
+    if (!elem(n).isInstanceOf[Tuple]) {
+      Left(TypeMismatch(n, Tuple.getClass().getName()))
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Should be `Tuple.getClass.getName`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package coop.rchain.rosette.prim
+
+import coop.rchain.rosette.{Ctxt, Ob, RblBool, Tuple}
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Still some unnecessary imports. IntelliJ can help you with that: `Code -> Optimize Imports`.

Comment (rabbitonweb) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: we might consider using WartRemover to have it done on compile time (stop compilation if unused imports in source code)

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Tuple] = {
+      val elem = ctxt.argvec.elem
+      checkTuple(1, elem).map(Tuple.cons(elem.head, _)) // arg1 must be a Tuple
+    }
+
+  }
+
+  private def checkTuple(n: Int, elem: Seq[Ob]): Either[PrimError, Tuple] =
```
Comment (rabbitonweb) [inline roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: I see a lot of usage of `Either`, maybe worth considering is `Disjunction` from Scalaz? Much richer API plus (more importantly) right-biased by default

Comment (tymm) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Since 2.12.x `Either` is right-biased by default.

Comment (henrytill) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Also, `cats.syntax.either` enriches the standard library `Either` API. 

(and makes it backward compatible with 2.11, though I doubt that matters in this project)

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala (hunk 10 above)
Comment (henrytill) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Huge wartremover fan here.

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala (hunk 11 above)
Comment (rabbitonweb) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: @tymm was not aware, cool!
@henrytill scalaz vs cats, I honestly don't care as long we use on or the other :)

Comment (henrytill) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: @rabbitonweb agreed!

`roscala` seems to be using cats.  i'm using cats in `storage`.  `rholang`, OTOH, is using `scalaz`, so...

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala (hunk 10 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Done

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala (hunk 7 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Done

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala (hunk 9 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Done.

File: roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala (hunk 8 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala]: Done.

**Comment by pyrocto** [comment]
Are you stuck on something?

**Comment by kjmcqueen** [comment]
Trying to get my test cases to compile.
